protocol/consistencychecks.go

```diff
@@ -104,84 +100,118 @@ func (cc *ConsistencyChecks) Verify(requestType int, msg *Response,
 	}
 
 	var verifResult error
-	var proofType int
-
 	cloneCC := cc.clone()
-	switch msg.DirectoryResponse.(type) {
-	case *DirectoryProof:
-		proofType, verifResult = cloneCC.verifyDirectoryProof(requestType,
-			msg.Error, msg.DirectoryResponse.(*DirectoryProof), uname, key)
 
-	case *DirectoryProofs:
-		proofType, verifResult = cloneCC.verifyDirectoryProofs(
-			msg.DirectoryResponse.(*DirectoryProofs), uname, key)
+	switch requestType {
+	case RegistrationType:
+		verifResult = cloneCC.verifyRegistration(requestType, msg, uname, key)
+	case KeyLookupType:
+		verifResult = cloneCC.verifyKeyLookup(requestType, msg, uname, key)
+	case MonitoringType:
+	case KeyLookupInEpochType:
+	default:
+		panic("[coniks] Unknown request type")
+	}
+
+	if verifResult == PassedWithAProofOfAbsence ||
+		verifResult == PassedWithAProofOfInclusion {
+		cc.update(cloneCC)
+	}
+
+	return verifResult.(ErrorCode)
+}
+
+func (cc *ConsistencyChecks) verifyRegistration(requestType int,
+	msg *Response, uname string, key []byte) error {
+
+	df, ok := msg.DirectoryResponse.(*DirectoryProof)
+	if !ok {
+		return ErrorMalformedDirectoryMessage
+	}
+
+	ap := df.AP
+	str := df.STR
+
+	// 1. verify STR
+	if verifResult := cc.verifySTR(str); verifResult != nil {
+		return verifResult
 	}
 
+	// 2. verify Auth path
+	proofType, verifResult := cc.verifyAuthPath(uname, key, ap, str)
 	if verifResult != nil {
-		return verifResult.(ErrorCode)
+		return verifResult
+	}
+
+	// 3. verify (response code, proof type) pair.
+	switch {
+	case msg.Error == ErrorNameExisted:
```

Comment:
There are two scenarios in which we get an `ErrorNameExisted`, right? We either have a proof of inclusion and no TB, or we have a TB and a proof of absence, right? I'm wondering if we should test for each of these scenarios individually at this point. I know `verifyReturnedPromise` ensures it's a proof of absence, but I would prefer to fail earlier if this isn't the case.

I'm also wondering if it may be worth checking for the expected (result, proof type) combination before even doing the auth path verification and fail early if something is wrong.